Allow switching sub-environment inside Manage Environment modal

cobwebsonsale opened the pull request:
<!--
Please open an [Issue](https://github.com/kong/insomnia/issues/new) first to discuss new
features or non-trivial changes. Please provide as much detail as possible on the change as
possible including general description, implementation details, potential shortcommings, etc.

If this PR closes an issue, please mention "Closes #XX" where #XX is the issue number.
-->
Closes #2761

- Added a button next to the sub-environment name to activate it
- Added a filled button next to the current active sub-environment
- Added helpful text in tooltips
- Renamed occurrences of `activeEnvironment` to `selectedEnvironment` in `WorkspaceEnvironmentsEditModal` (to avoid confusion) -- separated in another commit, for easier review/revert.

Caveat:
Could not re-render the currently selected environment in the right-pane after activation, so showing the activated environment instead. The re-render only makes sense when Base Environment is selected, as a different active environment changes the validations on variables. Tried:
- calling `_load`, 
- adding `activeEnvironmentId` to `state` inside `WorkspaceEnvironmentsEditModal` and setting state, and
- deriving `activeEnvironmentId` from `activeWorkspaceMeta` in `Wrapper`, as that is what gets updated in `handleChangeEnvironment`

Demo:
![switch-environment](https://user-images.githubusercontent.com/1848127/100902156-a333ba80-34ea-11eb-8c12-d6b27bf0c343.gif)

develohpanda commented:
Wow, thanks a lot for tackling this @cobwebsonsale! 🤗

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/insomnia?pullRequest=2891) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/insomnia?pullRequest=2891) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/insomnia?pullRequest=2891) it.</sub>

cobwebsonsale commented:
Requesting a review here.

develohpanda commented:
I'm sorry about the mega delays @cobwebsonsale 😞 I'll try to at least give a prelim review this this week

develohpanda reviewed (approved):
Thank you for the PR, QA'd locally and seems to be working as intended, and thanks for splitting up your work into individual commits which made the rename and feature easier to review. 🙌🏽 

My apologies for the delays. Unfortunately the team is spread quite thin at the moment, evident by the sheer number of open-source PRs we need to get through. We currently have a code-freeze in place for migrating to typescript, but once that is done I'll come back to this PR to merge it.

develohpanda commented:
Sounds good 👍🏽

This is a note to myself to double-check the caveat you describe above and whether that's a concern right now; with the testing done so far it seemed to be fine

cobwebsonsale committed:
Add ability to switch env inside manage env modal

- Show env status next to name
- Add tooltip for statuses
- Show env after activation

dimitropoulos committed:
TypeScript updates

note the end of line 5.  that was the reason none of these types were used before.

dimitropoulos commented:
I updated this branch for TypeScript: it seems to be working well - 

https://github.com/dimitropoulos/insomnia/commits/feature/switch-env-ts

The first commit is yours (please let me know if I messed up anything on the revival) and the second commit is to fix some TypeScript errors now that we're actually using types for `react-sortable-hoc`.

Please take a look.  If everything is ok, I can push that branch to this one and update this PR, at which point we should be ready (or nearly ready) to go.

develohpanda commented:
@dimitropoulos having [`2286466` (#2891)](https://github.com/Kong/insomnia/pull/2891/commits/2286466663fec0b4269162a52d24157aba1580ae) and [`d1c4ae2` (#2891)](https://github.com/Kong/insomnia/pull/2891/commits/d1c4ae2a5d2c039c8446128405bdf4f4e1a3bd3d) separate certainly was helpful in processing this PR, but if that's tricky to retain it's okay. I briefly checked your branch and https://github.com/dimitropoulos/insomnia/commit/c5eff8a791a95a24cb00613a9e246eda9889b46b looks correct.

Please push those changes up and we'll run through another code-review and QA round.

cobwebsonsale committed:
Rename activeEnvironment to selectedEnvironment in env edit modal

dimitropoulos commented:
@develohpanda sure, I un-squashed the rename and pushed to the PR's branch.

develohpanda commented on packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.tsx:
```diff
@@ -79,22 +93,44 @@ const SidebarListItem = SortableElement(
 
-const SidebarList = SortableContainer(
-  ({ environments, activeEnvironment, showEnvironment, changeEnvironmentName }) => (
+interface SidebarListProps extends Omit<SidebarListItemProps, 'environment'> {
```
Nice use of `Omit`!

develohpanda commented on packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.tsx:
```diff
@@ -225,7 +261,7 @@ class WorkspaceEnvironmentsEditModal extends PureComponent<Props, State> {
     // Unset active environment if it's being deleted
     if (activeEnvironmentId === environment._id) {
-      await handleChangeEnvironment(null);
+      handleChangeEnvironment(null);
```
Removal of this await actually shows a bug where `Wrapper._handleChangeEnvironment` function swallows a promise. It should be awaited correctly so we don't proceed to delete an environment until the change is complete.

Can be handled separate to this PR

develohpanda reviewed (approved):
Alright, double checked again and it looks good to me

> Could not re-render the currently selected environment in the right-pane after activation, so showing the activated environment instead. The re-render only makes sense when Base Environment is selected, as a different active environment changes the validations on variables.

I worked through this caveat you described; it's interesting that the re-render doesn't occur but when hovering over a template tag in the base environment the sub environment in that tooltip _does_ change but the resolved value does not. This leads me to believe there must be a bug somewhere and that it should work automatically.

Regardless, I still think what you have is fine for this iteration. :tada:

develohpanda committed:
update with develop

reynolek reviewed (approved):
Product perspective: Good addition, once merged we should probably do a design pass but that doesnt have to happen right now.
